packages/devtools/src/schema/compiler.ts

```diff
@@ -0,0 +1,315 @@
+import type { ModelDefinition, ModelViewsDefinition } from '@ceramicnetwork/stream-model'
+import type { JSONSchema } from '@composedb/types'
+import type { SetRequired } from 'type-fest'
+
+import type { AnySchema, ScalarSchema } from '../types.js'
+
+import { parseSchema } from './parser.js'
+import type {
+  AbstractModelDefinition,
+  ListFieldDefinition,
+  ObjectDefinition,
+  ObjectReferenceFieldDefinition,
+  ScalarFieldDefinition,
+  SchemaDefinition,
+} from './types.js'
+
+type CompileObject = SetRequired<JSONSchema.Object, 'properties'>
+
+type JSONSchemaReference = { $ref: string }
+
+type ReferencedSchema = ScalarSchema | JSONSchema.Object
```

Comment:
I see the comment below saying that scalars with title can be referenced, but could you add something similar here, just in case somebody wants to understand this type without going through the implementation?

Reply:
Basically any JSON Schema type that can be reused in a given model is created as a single definition with one or more reference(s).